src/storage/table/row_group.cpp

```diff
@@ -648,9 +676,125 @@ void RowGroup::MergeStatistics(idx_t column_idx, BaseStatistics &other) {
+	}
+	auto &config = DBConfig::GetConfig(db);
+	if (config.force_compression != CompressionType::COMPRESSION_AUTO) {
+		return config.force_compression;
+	}
+	if (!CompressionForTypeExists(columns[col_idx]->type.InternalType())) {
+		return CompressionType::COMPRESSION_AUTO;
+	}
+	vector<CompressionFunction *> compression_functions =
+	    db.config.GetCompressionFunctions(columns[col_idx]->type.InternalType());
+	D_ASSERT(!compression_functions.empty());
+
+	// set up the analyze states for each compression method
+	vector<unique_ptr<AnalyzeState>> analyze_states;
+	analyze_states.reserve(compression_functions.size());
+	for (idx_t i = 0; i < compression_functions.size(); i++) {
+		if (!compression_functions[i]) {
+			analyze_states.push_back(nullptr);
+			continue;
+		}
+		analyze_states.push_back(
+		    compression_functions[i]->init_analyze(*columns[col_idx], columns[col_idx]->type.InternalType()));
+	}
+	Vector intermediate(columns[col_idx]->type, true, false);
+	// scan over all the segments and run the analyze step
+	ScanSegments(
+	    [&](Vector &scan_vector, idx_t count) {
+		    for (idx_t i = 0; i < compression_functions.size(); i++) {
+			    if (!compression_functions[i]) {
+				    continue;
+			    }
+			    auto success = compression_functions[i]->analyze(*analyze_states[i], scan_vector, count);
+			    if (!success) {
+				    // could not use this compression function on this data set
+				    // erase it
+				    compression_functions[i] = nullptr;
+				    analyze_states[i].reset();
+			    }
+		    }
+	    },
+	    intermediate, col_idx);
+
+	// now that we have passed over all the data, we need to figure out the best method
+	// we do this using the final_analyze method
+	unique_ptr<AnalyzeState> state;
+	compression_idx = DConstants::INVALID_INDEX;
+	idx_t best_score = NumericLimits<idx_t>::Maximum();
+	for (idx_t i = 0; i < compression_functions.size(); i++) {
+		if (!compression_functions[i]) {
+			continue;
+		}
+		auto score = compression_functions[i]->final_analyze(*analyze_states[i]);
+		if (score < best_score) {
+			compression_idx = i;
+			best_score = score;
+			state = move(analyze_states[i]);
+		}
+	}
+	//	return state;
+	return compression_functions[compression_idx]->type;
+}
+
+vector<CompressionType> RowGroup::DetectBestCompressionMethodTable(TableDataWriter &writer) {
+	vector<CompressionType> table_compression;
+	for (idx_t i = 0; i < columns.size(); i++) {
+		idx_t compression_idx = 0;
+		CompressionType compression_type = writer.GetColumnCompressionType(i);
+		table_compression.push_back(DetectBestCompressionMethod(compression_idx, i, compression_type));
+	}
+	return table_compression;
+}
+
+RowGroupPointer RowGroup::Checkpoint(TableDataWriter &writer, vector<unique_ptr<BaseStatistics>> &global_stats,
+                                     DataTable &data_table) {
 	vector<unique_ptr<ColumnCheckpointState>> states;
 	states.reserve(columns.size());
+	if (db.config.force_compression_sorting) {
```

Comment:
Can we use the `config.preserve_insertion_order` setting to trigger this instead?

Reply:
I've added that to the if statement. Perhaps we should leave the `config.force_compression_sorting` as well, in case it should be explicitly turned off (e.g. for debugging)?

Reply:
One issue might be that `config.preserve_insertion_order` seems to be true for the unit tests?

Reply:
Perhaps disable it specifically for the tests you added?

Reply:
Done.